In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression

In [2]:
X , y = make_regression(n_samples=1000 , n_features=3 , noise=10 , random_state=42)

In [3]:
print(X.shape)

(1000, 3)


In [4]:
print(y.shape)

(1000,)


In [5]:
X = np.c_[np.ones(X.shape[0]), X]

In [6]:
y = y.reshape(len(y),1)

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
def calculate_loss(X, y, weights):
    
# # Example data
# X = np.array([[1, 2], [3, 4], [5, 6]])  
# y = np.array([5, 11, 17])              
# weights = np.array([1, 1])             

    #  Calculate predictions by multiplying inputs (X) with weights:)
    predictions = X.dot(weights)    #predictions=[1⋅1+2⋅1,3⋅1+4⋅1,5⋅1+6⋅1]=[3,7,11]
    
    #  Calculate the errors (difference between predictions and actual values):)
    errors = predictions - y  #errors=[3−5,7−11,11−17]=[−2,−4,−6]
    
    # Calculate Mean Squared Error (MSE):)
    mse = np.mean(errors ** 2)   #mse=error **2/the number of samples  
    
    #  Divide by 2 to match the gradient calculation format:)
    loss = mse / 2  
    
    return loss

Gradient Descent

In [9]:
def gradient_descent(X, y, num_iterations , freq_print):
    alpha = 0.001
    W = np.random.randn(X.shape[1], 1)
    for i in range(num_iterations) :    
        yhat = X.dot(W)
        error = yhat - y
        gradient = X.T.dot(error)
        W = W - alpha * gradient / len(X)
        if i % freq_print == 0:
            print(calculate_loss(X , y , W))
        
        
    return W             

In [10]:
W = gradient_descent(X_test , y_test , 10000 , 1000)

8312.655351815809
1221.2065884941326
229.73373299063024
79.16097386019601
54.43521626187781
50.08818622723116
49.28080884222707
49.12457542245187
49.09345644672667
49.087135838617215


In [11]:
W 

array([[ 1.6079116 ],
       [98.19389642],
       [81.983694  ],
       [25.92508874]])

## To test the gradient descent

this function give the most efficient weight and I print the loss for this weight

In [12]:
def least_weight(X, y):
    X_transpose = X.T
    W = np.linalg.inv(X_transpose @ X) @ X_transpose @ y
    
    
    return W
    

### print the loss of the most efficient weights

In [13]:
print(calculate_loss(X_test,y_test,least_weight(X_test,y_test)))

49.08549457182744


### print the loss of Gradient descent

In [14]:
print(calculate_loss(X_test,y_test,W))

49.08583601163943


#### The efficient weights loss is : 49.08549457182744
#### The gradient descent loss is: 49.08583601163943


Predicting the output for the test data

In [15]:
def predict(W, X):
    return X.dot(W)

In [16]:
predictions = predict(W, X_test)

In [17]:
comp = np.concatenate((predictions.reshape(len(predictions), 1), y_test), axis=1)


In [18]:
np.set_printoptions(precision=3 ,suppress=True)
comp

array([[ -74.848,  -82.293],
       [ 112.993,  104.808],
       [  51.101,   55.153],
       [ 180.673,  175.842],
       [ -22.549,  -12.822],
       [-162.556, -158.697],
       [ 134.344,  117.792],
       [  69.208,   64.592],
       [ 231.856,  236.283],
       [-112.752, -109.113],
       [-106.859, -103.8  ],
       [ 118.704,  114.629],
       [  31.358,   31.569],
       [ -15.916,    2.217],
       [  38.189,   43.259],
       [ -49.342,  -45.865],
       [-190.627, -181.071],
       [  61.96 ,   75.81 ],
       [  -7.789,   -6.023],
       [  43.73 ,   26.268],
       [ 102.933,  101.897],
       [  38.603,   44.527],
       [-127.62 , -125.496],
       [  15.838,   26.333],
       [-171.955, -173.389],
       [  92.482,   77.165],
       [   3.043,   -4.217],
       [ -72.397,  -78.773],
       [ -64.476,  -75.655],
       [ -48.47 ,  -40.12 ],
       [ 135.678,  128.33 ],
       [ -56.544,  -49.205],
       [  77.807,   77.515],
       [-300.458, -308.578],
       [  35.7

Checking the performance of the model


In [19]:
def mean_squared_error(y, y_pred):
    mse = (y - y_pred) ** 2
    return np.mean(mse)


In [20]:
mean_squared_error(y_test, predictions)

np.float64(98.17167202327886)

In [21]:
def mean_absolute_error(y, y_pred):
    mae = np.abs(y - y_pred)
    return np.mean(mae)

In [22]:
mean_absolute_error(y_test, predictions)

np.float64(8.058159173599313)